### ISSUE TITLE
Allows Certbot for Windows to run without administrative privileges

### ISSUE THREAD

**Issue by adferrand** [certbot/certbot#7881]
Currently Windows is installed with administrative privileges and needs to be run from an elevated prompt to run.

This is mainly due to the creation of symlinks during Certbot execution. This topic is handled in #7880. Once done, we can modify the remaining logic in Certbot to run it with a normal user.

**Comment by bmw** [comment]
> This is mainly due to the creation of symlinks during Certbot execution.

Is this true? Does Certbot need admin privileges on Windows to:

* Bind to port 80/443 when using the standalone plugin
* Modify the Apache/nginx/IIS config
* Modify the certificate store
* Modify the default "webroot" of Apache/nginx/IIS

We need root do these things on Linux in almost all configurations.

If we don't (usually) need admin privileges, then I think resolving issues like #7880 and https://github.com/certbot/certbot/issues/7883 could be pretty interesting.

If we do (usually) need admin privileges to do these things though, especially if it is needed to have Certbot install the certificate for you which is what we usually encourage people do, then I'm personally less motivated to make these changes. Ideally, we should always use the fewest permissions possible, but practically I honestly think it introduces complexity and potential problems both for us as maintainers of our code and for our users about potential things that can go wrong. Failing fast with a simple statement that you need admin privileges simplifies things a lot I think.

That's not to say that if we do need admin privileges for these things we can't consider some of this stuff, however, if its main benefit is improving cases like `certbot certonly` with a DNS plugin, I'm personally not very motivated to prioritize this work as I don't think its the strength of Certbot compared to other ACME clients.

Finally, if you're largely motivated to solve these issues for dnsrobocert, we can maybe talk about ways to make it easy for you to programmatically achieve the behavior you want even if it's still maybe hard to do for an average user using Certbot directly.

What do you think?

**Comment by adferrand** [comment]
Let's precise my point here.

This issue is not about if Certbot needs root/administrative privileges or not to work properly. It is about that Certbot *requires* administrative privileges proactively on Windows while it does not on Linux.

Yes by default, most basic services like binding 80/443 or manage IIS require these privileges. However you can give access to specific users to whole or part of IIS, and you can have user certificates stores.

Anyway the point is that depending on the specific configuration of a machine, Certbot can work or not properly without high privileges. It should not be up to Certbot to check that, but to the user.

So I would like Certbot for Windows to behave like on Linux, and not require administrative privileges to start. Up to the user to check if it will work or not depending on the plugin used.

Currently it makes this check because of the symlinks privileges. I would like to remove that.

Common usecase on IIS here: this HTTP servers allows you to create per-user VirtualHost equivalent (it is called a site on IIS). You control its configuration: Certbot should be able to run as a normal user to protect this specific site with HTTPS.

NB: As I said live, this issue does not concern the DNS plugins or DNSroboCert (even if it would help, obviously), but more generally how Certbot can be used on Windows.

**Comment by bmw** [comment]
One thing we could probably do pretty easily if you wanted is to change Certbot's explicit check for admin privileges on Windows to a check for the ability to create symlinks. I think we'd probably want to continue to do this early in execution as doing it later has the potential to crash when trying to create the lineage which I think should be avoided, but it would remove Certbot's unconditional requirement of admin privileges.

If you want to remove the need to create symlinks entirely though as proposed in #7880, I'm honestly pretty hesitant. With the way I'd initially like to approach the issue which I've described in #7880, I think it is a major project and while it certainly has value, it doesn't seem to me as impactful as other issues we could work on such as snaps, DNS plugins, 3rd party/DNS plugins on Windows, or even an IIS plugin. Let me know if you disagree with this, but otherwise, I think we should postpone this discussion to when/if we think it's the best use of our time to do this work.

**Comment by adferrand** [comment]
Sadly testing for the symlinks creation capability is equivalent to test for administrative access, since only Administrators have this capability by default.

In general I think this precise issue, requirement of administrative privileges, is a major issue, that should be fixed before any other further functionalities we could add to Certbot on Windows, and a blocker to move Certbot out of the beta phase.

Let me explain. Biggest problems are usability and security.

In term of usability, requiring administrative privileges prevent normal users to install Certbot. For various cases like IIS, this reduces a lot the usability, while Windows is designed to run a lot of programs in user space.

But the biggest problem is the security. Here Certbot has administrative privileges. Any flaws that we could introduce in Certbot codebase, or any developer in libraries Certbot depends on, could be a backdoor to execute arbitrary code with privileges permissions. However we do not even need to think of these flaws, it is already a problem with Certbot features: here the core point is the ability to run hooks (either from command line or in the proper hook folder) that would be run during the certificate lifecycle. An attacker may temper the existing hook scripts, or add new ones, that will be executed by Certbot ...

I know it is already a problem on Linux when Certbot is run as root. But because of Windows specificites, and its surface attack this is way broader compared to Linux, it becomes critical for this OS.

So I think we need to find a way to get rid of these administrative privileges. Most Windows administrators would refuse to install Certbot if is not possible to do it with a specific user designed for a specific purpose (like a user handling a particular IIS site).

If you agree with me at this point, I can provide some high level design propositions. It is mostly about removing (like explained in #7880), or modifying symlinks. I have also another idea in mind that would occur at the installation phase for Windows only.

What do you think of that?

**Comment by bmw** [comment]
I know only administrators have the ability to create symlinks by default, but my understanding is it can be granted to other users, at least in newer versions of Windows. Changing Certbot's code to check for the ability to create symlinks would provide a workaround for those who only need admin permissions to create symlinks. I know this doesn't fully solve the issue in the way that you want, but it still seems to me like an easy way to improve things if it interests you.

I'm having a hard time accepting your overall argument though. First, we agreed earlier in this thread that most Windows users will need to run Certbot with admin privileges in order for Certbot to do what it needs like bind to port 80/443 or manage IIS by default. Second, we currently have over 1600 Windows users and I have yet to see a single request for this feature. While I agree that it is possible to create configurations where Certbot does not need admin privileges on Windows, it doesn't yet seem to me that these scenarios or wanting to use Certbot in them is common enough for us to spend any time on this right now.

For the security argument, it's not clear to me how the situation is any different than on Linux. In either scenario, if an attacker can get Certbot to execute their code, they can do whatever they want to the system. Certbot hooks should also have the same protections on Windows as they do on Linux. Am I missing something here?

In short, I'm still not convinced that this is the most impactful use of our time right now. With that said, if this is what you want to work on, you're welcome to provide some high level designs and we can try to consider them in the future, but until I see stronger reasons to make these changes, we honestly probably wouldn't find/make time to review them for quite a while.

**Comment by evilensky** [comment]
Certbot user here.  First apologies if issues are not the place of discussion like this, if so please delete this comment.

> earlier in this thread that most Windows users will need to run Certbot with admin privileges in order for Certbot to do what it needs like bind to port 80/443 or manage IIS by default.

This seems to contradict the [FAQ Entry](https://certbot.eff.org/faq/#does-certbot-require-root-administrator-privileges).

If one was to replace the [Raise-when-not-admin](https://github.com/certbot/certbot/blob/master/certbot/certbot/compat/misc.py#L30-L40) to focus on just the symlink privilege, what is the preferred style?  Try-to-make-symlink and Except or explicit check for the presence of the privilege?  Could it rely simply on the existing exception handling around symlink creation?

**Comment by eidylon** [comment]
We have a scheduled task which originally, the plan was for it to generate any new certs needed, and zip them up into a package ready to be delivered to our hosting company for installation into the load balancers. (They don't support win-acme, already asked.) 

So I have a program which gets the sites from IIS, checks their cert expiry, and - if needed - calls certbot for manual cert generation for the site, dumping the output into the C:\Certbot\live directory. Certbot is not directly accessing or updating IIS, or needing to register any ports, I don't think. 

The whole process works beautifully when I'm logged in with my account. 
However the scheduled task we have which runs under our service account fails, and the StdErr when calling certbot shows "Error, certbot must be run on a shell with administrative rights."

We're calling certbot from C# code, using Process.Start(). There are ways to elevate, but it will cause UAC prompts, which obviously doesn't work for a nightly scheduled task. 

The service account is the owner of the C:\Certbot directory, and as such has full R/W rights on the whole tree downward, and the scheduled task is set to "Run with highest privileges". 

Copied over here as suggested on my original LE forum post... 
https://community.letsencrypt.org/t/no-output-from-certbot-under-scheduled-task-due-to-admin-rights/160245 .

**Comment by bmw** [comment]
We're not planning on expanding our Windows support at this time so I'm going to close this for now.
